rpc/ethereum/namespaces/debug/api.go

```diff
@@ -131,87 +169,70 @@ func (a *API) TraceBlockByNumber(height rpctypes.BlockNumber, config *evmtypes.T
 		return nil, err
 	}
 
-	return a.traceBlock(height, config, resBlock.Block.Txs)
+	return a.traceBlock(height, config, resBlock)
 }
 
 // traceBlock configures a new tracer according to the provided configuration, and
 // executes all the transactions contained within. The return value will be one item
 // per transaction, dependent on the requested tracer.
-func (a API) traceBlock(height rpctypes.BlockNumber, config *evmtypes.TraceConfig, txs types.Txs) ([]*evmtypes.TxTraceResult, error) {
+func (a API) traceBlock(height rpctypes.BlockNumber, config *evmtypes.TraceConfig, block *tmrpctypes.ResultBlock) ([]*evmtypes.TxTraceResult, error) {
+	txs := block.Block.Txs
 	txsLength := len(txs)
 
 	if txsLength == 0 {
 		// If there are no transactions return empty array
 		return []*evmtypes.TxTraceResult{}, nil
 	}
 
-	var (
-		results = make([]*evmtypes.TxTraceResult, txsLength)
-		wg      = new(sync.WaitGroup)
-		jobs    = make(chan *evmtypes.TxTraceTask, txsLength)
-	)
+	txsMessages := make([]*evmtypes.MsgEthereumTx, txsLength)
+	txDecoder := a.clientCtx.TxConfig.TxDecoder()
+	for i, tx := range txs {
+		decodedTx, err := txDecoder(tx)
+		if err != nil {
+			a.logger.Error("failed to decode transaction", "hash", txs[i].Hash(), "error", err.Error())
+			continue
+		}
 
-	threads := runtime.NumCPU()
-	if threads > txsLength {
-		threads = txsLength
+		messages := decodedTx.GetMsgs()
+		if len(messages) == 0 {
+			continue
+		}
+		ethMessage, ok := messages[0].(*evmtypes.MsgEthereumTx)
+		if !ok {
+			// Just considers Ethereum transactions
+			continue
+		}
+		txsMessages[i] = ethMessage
```

Comment:
see msg above, use `append`
```suggestion
		txsMessages = append(txsMessages, ethMessage)
```